polkadot/node/network/collator-protocol/src/validator_side_experimental/peer_manager/mod.rs

```diff
@@ -0,0 +1,495 @@
+	) -> Result<()> {
+		for leaf in new_leaves {
+			let block_number = get_block_number(sender, leaf).await?;
+
+			let latest_processed_block_number = self.db.latest_block_number().await;
+			if latest_processed_block_number.unwrap_or(0) > block_number {
+				continue
+			}
+
+			let parent = if let Some(parent) = cached_parents.get(&leaf) {
+				*parent
+			} else {
+				let parent = get_ancestors(sender, 1, leaf).await?.remove(0);
+				cached_parents.insert(leaf, parent);
+				parent
+			};
+
+			let ancestors = vec![parent, leaf];
+			let bumps = extract_reputation_bumps_from_ancestry(sender, &ancestors[..]).await?;
+
+			let updates = self
+				.db
+				.process_bumps(
+					block_number,
+					bumps,
+					Some(Score::new(INACTIVITY_DECAY).expect("INACTIVITY_DECAY is a valid score")),
+				)
+				.await;
+			for update in updates {
+				self.connected.update_reputation(update);
+			}
+		}
+
+		Ok(())
+	}
+
+	/// Process the registered paras and cleanup all data pertaining to any unregistered paras, if
+	/// any. Should be called every N finalized block notifications, since it's expected that para
+	/// deregistrations are rare.
+	pub async fn registered_paras_update(&mut self, registered_paras: BTreeSet<ParaId>) {
+		// Tell the DB to cleanup paras that are no longer registered. No need to clean up the
+		// connected peers state, since it will get automatically cleaned up as the claim queue
+		// gets rid of these stale assignments.
+		self.db.prune_paras(registered_paras).await;
+	}
+
+	/// Process a potential change of the scheduled paras.
+	pub async fn scheduled_paras_update<Sender: CollatorProtocolSenderTrait>(
+		&mut self,
+		sender: &mut Sender,
+		scheduled_paras: BTreeSet<ParaId>,
+	) {
+		let mut prev_paras_count = 0;
+		let mut prev_scheduled_paras = self.connected.scheduled_paras();
+
+		if prev_scheduled_paras.all(|p| {
+			prev_paras_count += 1;
+			scheduled_paras.contains(p)
+		}) {
+			// The new set is a superset of the old paras and their lengths are equal, so they are
+			// identical.
+
+			if prev_paras_count == scheduled_paras.len() {
+				// Nothing to do if the scheduled paras didn't change.
+				return
+			}
+		}
+
+		// Borrow checker can't tell that prev_scheduled_paras is not used anymore.
+		std::mem::drop(prev_scheduled_paras);
+
+		// Recreate the connected peers based on the new schedule and try populating it again based
+		// on their reputations. Disconnect any peers that couldn't be kept
+		let mut new_instance =
+			ConnectedPeers::new(scheduled_paras, CONNECTED_PEERS_LIMIT, CONNECTED_PEERS_PARA_LIMIT);
+
+		std::mem::swap(&mut new_instance, &mut self.connected);
+		let prev_instance = new_instance;
+		let (prev_peers, cached_scores) = prev_instance.consume();
+
+		// Build a closure that can be used to first query the in-memory past reputations of the
+		// peers before reaching for the DB.
+		// TODO: we could warm-up the DB for these specific paraids. (Cache them in memory on the
```

Comment:
The DB uses the OS page cache, so likely not much improvement here.

Reply:
We'll see how it goes. I removed this todo as it's not the right time for it now